python/cudf_polars/cudf_polars/dsl/expressions/base.py

```diff
@@ -32,100 +33,17 @@ class ExecutionContext(IntEnum):
     ROLLING = enum.auto()
 
 
-class Expr:
-    """
-    An abstract expression object.
+class Expr(Node):
+    """An abstract expression object."""
 
-    This contains a (potentially empty) tuple of child expressions,
-    along with non-child data. For uniform reconstruction and
-    implementation of hashing and equality schemes, child classes need
-    to provide a certain amount of metadata when they are defined.
-    Specifically, the ``_non_child`` attribute must list, in-order,
-    the names of the slots that are passed to the constructor. The
-    constructor must take arguments in the order ``(*_non_child,
-    *children).``
-    """
-
-    __slots__ = ("dtype", "_hash_value", "_repr_value")
+    __slots__ = ("dtype",)
     dtype: plc.DataType
     """Data type of the expression."""
-    _hash_value: int
-    """Caching slot for the hash of the expression."""
-    _repr_value: str
-    """Caching slot for repr of the expression."""
     children: tuple[Expr, ...] = ()
     """Children of the expression."""
     _non_child: ClassVar[tuple[str, ...]] = ("dtype",)
     """Names of non-child data (not Exprs) for reconstruction."""
 
-    # Constructor must take arguments in order (*_non_child, *children)
-    def __init__(self, dtype: plc.DataType) -> None:
-        self.dtype = dtype
```

Comment:
Why remove this? It seems like every child class is setting this anyway now, right? Wouldn't it make more sense to continue having the `super` calls than setting the dtype manually in each case (in case we ever have to do anything different)?

Reply:
Perhaps, but I don't really like inheritance, so I'm kind of (ab)using the class hierarchy for algebraic datatypes.

Also, not calling `super` and just setting the properties directly is faster (about 25% for trivial classes)